doc/pins/batch_requests_api_hld.md

```diff
@@ -0,0 +1,184 @@
+
+### Batched APIs
+
+The following are the proposed batched APIs:
+
+```
+    virtual void set(const std::vector<KeyOpFieldsValuesTuple> &values) override;
+
+    virtual void del(const std::vector<std::string> &keys) override;
+```
+
+The new API will be added into the ProducerStateTable as new methods. The current set() and del() method will not be impacted.
+
+### Batched API Operations
+
+The Redis DB operation sequence will be the same as the current sequence as calling the current API multiple times, except sending the notification once at the end. There is no change in the consumer API and behavior. (Refer to the appendix for the current API operation.)
+
+![drawing](images/batch_requests_api_hld_batch_operation.png)
+
+The sequence of operations are:
+
+* Inserts all new keys into a Redis set "&lt;table_name>_KEY_SET".
+* For del() operation, also inserts the keys into another set "&lt;table_name>_DEL_SET".
+* Inserts the entries into a temporary Redis hash "_&lt;table_name>&lt;separator>&lt;key>".
+* After all entries are written, write a message "G" to the notification channel "&lt;table_name>_CHANNEL". This will notify the consumer.
+* All the above Redis operations will happen in a Redis transaction (LUA script), which is atomic.
+
+The following sequence diagram shows the detailed batched operation in the ProducerStateTable. The Redis new key set and the temporary hash are abstracted as the "APPL DB temp table" in the graph.
+
+![drawing](images/batch_requests_api_hld_sequence_diagram.png)
+
+### Orchagent Update
+
+The current consumer implementation is capable of handling batches. The batch size is configurable in the orchagent argument. A recent change in orchagent makes orchagent to pop in a loop until all requests are drained. This handles the case that the producer writes a batch larger than the orchagent pop size. Overall, there is no change required in the orchagent.
+
+### PINS P4RT
+
+The P4RT application will be the first application to use the batch APIs. P4Orch route manager will utilize the batching to maximize the performance. To handle APPL DB updates, we are proposing the batched API in the Table class in a similar fashion to the ProducerStateTable:
+
+```
+    virtual void set(const std::vector<KeyOpFieldsValuesTuple> &values);
+
+    virtual void del(const std::vector<std::string> &keys);
+```
+
+## SAI API
+
+N/A
+
+## Configuration and management
+
+N/A
+
+## Warmboot and Fastboot Design Impact
+
+N/A
+
+## Restrictions/Limitations
+
+* In the current proposed APIs, a batched request can only contain similar entries - for example - either all SET or all DEL entries. A mix of SET and DEL in a batch is not supported.
+* A batched request can only apply to a single APPL DB table. Requests to different tables cannot be batched.
+* Order is not maintained in a batch request. The entries within a single batched request will be handled in any order.
+
```

Comment:
Would this not cause a problem for SET case, if order is not maintained for SET, two sets on same object can result in random state at the end.

Reply:
Two sets of the same object will have the final result of the combined attributes.

Order-less is the limitation of the current ProducerStateTable APIs, for both batch and non-batch APIs. Even in the non-batch APIs, if the producer writes multiple requests in a short time, the consumer might pop multiple entries at the time it wakes up. The reason of the order-less is that the API uses a Redis set to store the new keys, which is order-less.